src/properties.jl

```diff
@@ -0,0 +1,668 @@
+end
+
+
+@propertyclass GenericProperties API.H5P_DEFAULT
+
+superclass(::Type{P}) where {P<:Properties} = GenericProperties
+
+class_getproperty(::Type{GenericProperties}, props, name) =
+    error("$(typeof(props)) has no property $name")
+class_setproperty!(::Type{GenericProperties}, props, name, val) =
+    error("$(typeof(props)) has no property $name")
+all_propertynames(::Type{GenericProperties}) = ()
+
+
+# for initializing multiple Properties from a set of keyword arguments
+"""
+    setproperties!(props::Properties...; kwargs...)
+
+For each `(key, value)` pair in `kwargs`, set the corresponding properties in
+each `Properties` object in `props`. Returns a `Dict` of any pairs which didn't
+match properties in `props`.
+"""
+function setproperties!(props::Properties...; kwargs...)
+    filter(kwargs) do (k,v)
+        found = false
+        for prop in props
+            if k in all_propertynames(typeof(prop))
+                setproperty!(prop,k,v)
+                found = true
+            end
+        end
+        return !found
+    end
+end
+
+
+
+###
+### Convenience macros for defining getter/setter functions
+###
+
+"""
+    @tuple_property(name)
+"""
+macro tuple_property(property)
+    get_property = Symbol(:get_,property)
+    set_property! = Symbol(:set_,property,:!)
+    api_get_property = :(API.$(Symbol(:h5p_get_,property)))
+    api_set_property = :(API.$(Symbol(:h5p_set_,property)))
+    quote
+        function $(esc(get_property))(p::Properties)
+            return $api_get_property(p)
+        end
+        function $(esc(set_property!))(p::Properties, val::Tuple)
+            return $api_set_property(p, val...)
+        end
+    end
+end
+
+"""
+    @enum_property(name, sym1 => enumvalue1, sym2 => enumvalue2, ...)
+
+Wrap property getter/setter API functions that use enum values to use symbol instead.
+"""
+macro enum_property(property, pairs...)
+    get_property = Symbol(:get_,property)
+    set_property! = Symbol(:set_,property,:!)
+    api_get_property = :(API.$(Symbol(:h5p_get_,property)))
+    api_set_property = :(API.$(Symbol(:h5p_set_,property)))
+
+    get_expr = :(error("Unknown $property value $enum"))
+    set_expr = :(throw(ArgumentError("Invalid $property $val")))
+
+    for pair in reverse(pairs)
+        @assert pair isa Expr && pair.head == :call && pair.args[1] == :(=>)
+        _, val, enum = pair.args
+        get_expr = :(enum == $enum ? $val : $get_expr)
+        set_expr = :(val == $val ? $enum : $set_expr)
+    end
+    quote
+        function $(esc(get_property))(p::Properties)
+            property = $(QuoteNode(property))
+            enum = $api_get_property(p)
+            return $get_expr
+        end
+        function $(esc(set_property!))(p::Properties, val)
+            property = $(QuoteNode(property))
+            enum = $set_expr
+            return $api_set_property(p, enum)
+        end
+        function $(esc(set_property!))(p::Properties, enum::Integer)
+            # deprecate?
```

Comment:
I agree we should deprecate this, but it seems tricky to do this since it requires the inverse dict mapping and calling the `@deprecate` macro within the quoted block of this return.

Reply:
Yeah, I couldn't see an easy way to do it.